### PR TITLE
feat(manager/poetry): pass POETRY_SOLVER_MIN_RELEASE_AGE when minimumReleaseAge is configured

### LINKED ISSUE #43430: Pass `POETRY_SOLVER_MIN_RELEASE_AGE` to Poetry when `minimumReleaseAge` is configured
Per noted in #43428, Poetry 2.4.0's [`solver.min-release-age`](https://python-poetry.org/docs/configuration/) should be specified (if not already set in the project) when Renovate's `minimumReleaseAge` applies to given update(s)

### PR DESCRIPTION
**Pull request by LuigiCerone**
## Changes

When `minimumReleaseAge` is configured in Renovate, Poetry invocations during lock file updates now set the `POETRY_SOLVER_MIN_RELEASE_AGE` environment variable so that Poetry's solver enforces the age constraint on all resolved packages, including transitive dependencies.

- Converts Renovate's `minimumReleaseAge` duration string to an integer number of days
- Sets `POETRY_SOLVER_MIN_RELEASE_AGE` when invoking `poetry update --lock`
- Applies to both regular dependency updates and `lockFileMaintenance` runs

Without this change, `minimumReleaseAge` only guards direct dependencies that Renovate proposes.
When Poetry regenerates `poetry.lock`, it can silently pull in brand-new transitive dependencies with no age filter.
Setting `POETRY_SOLVER_MIN_RELEASE_AGE` closes that gap.

Notes:
- Requires Poetry `>=2.4.0`
- `solver.min-release-age` is a global Poetry config (set via `poetry config` or env var), not a `pyproject.toml` setting

## Context

- [x] This closes an existing Issue, Closes: #43430 (sub-feature of #41652, discussed in #43428)

## AI assistance disclosure

- [x] Yes — substantive assistance (AI-generated non‑trivial portions of code, tests, or documentation). Used Claude Sonnet 4.6 for implementation, test writing, and documentation.

## Documentation (please check one with an [x])

- [x] I have updated the documentation

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Newly added/modified unit tests, or

The public repository: https://github.com/LuigiCerone/renovate-min-repro-template

### CODE REVIEW

**Review by jamietanna — CHANGES_REQUESTED** [review]
Please fill in our PR template + address CI failures

This will require a minimal reproduction repo to validate the changes work as expected

**Comment by jamietanna** [comment]
Also, please avoid force pushing after we've started reviewing, as it makes it harder to see changes

**Comment by LuigiCerone** [comment]
Hello @jamietanna , thanks for the guidance. I've tried to address all the requests, LMK. Thanks!

File: docs/usage/key-concepts/minimum-release-age.md [hunk 1]
```diff
@@ -60,6 +60,14 @@ This ensures existing lock files are never broken by the `--before` flag.
 
 After the next lock file maintenance run (which regenerates the lock file from scratch with `--before`), subsequent updates will fully enforce the `minimumReleaseAge` constraint.
 
+#### Poetry
+
+When `minimumReleaseAge` is configured, Renovate sets the `POETRY_SOLVER_MIN_RELEASE_AGE` environment variable when running `poetry lock`.
+This ensures that Poetry only resolves package versions that are at least as old as the cooldown threshold.
```
Comment (RahulGautamSingh) [inline docs/usage/key-concepts/minimum-release-age.md:79]: This says `poetry lock` but the code runs `poetry update --lock`

File: docs/usage/key-concepts/minimum-release-age.md [hunk 2]
```diff
@@ -60,6 +60,14 @@ This ensures existing lock files are never broken by the `--before` flag.
...
+
+The value is calculated as `ceil(minimumReleaseAge / 1 day)` and passed as an integer number of days.
+If `solver.min-release-age` is already set in `[tool.poetry.solver]` in `pyproject.toml`, Renovate uses the stricter (larger) of the two values.
+
 ### What happens if the datasource and/or registry does not provide a release timestamp, when using `minimumReleaseAge`?
```
Comment (RahulGautamSingh) [inline docs/usage/key-concepts/minimum-release-age.md:83]: The code doesn't check `pyproject.toml` at all. It just sets the env var unconditionally. The PR description itself says this is a global Poetry config, not a `pyproject.toml` setting. 

Implement the said behavior or remove this line.

File: lib/modules/manager/poetry/artifacts.ts [hunk 3]
```diff
@@ -224,6 +231,16 @@ export async function updateArtifacts({
       PIP_CACHE_DIR: await ensureCacheDir('pip'),
     };
 
+    if (config.minimumReleaseAge) {
+      const ageMs = toMs(config.minimumReleaseAge);
+      if (isNumber(ageMs)) {
+        const now = DateTime.now();
+        const cutoff = now.minus(ageMs).toUTC();
```
Comment (RahulGautamSingh) [inline lib/modules/manager/poetry/artifacts.ts]: This is just `Math.ceil(ageMs / 86_400_000)`. The `DateTime` import + `.toUTC()` + diff  adds complexity for no reason.

File: lib/modules/manager/poetry/artifacts.spec.ts [hunk 4]
```diff
@@ -699,5 +699,74 @@ describe('modules/manager/poetry/artifacts', () => {
...
+          options: {
+            env: expect.not.objectContaining({
+              POETRY_SOLVER_MIN_RELEASE_AGE: expect.anything(),
+            }),
+          },
+        },
+      ]);
+    });
   });
 });
```
Comment (RahulGautamSingh) [inline lib/modules/manager/poetry/artifacts.spec.ts:822]: Missing edge case tests: sub-day durations (e.g. `"12 hours"` → `1`) and `"0 days"`

File: lib/modules/manager/poetry/artifacts.ts [hunk 5]
```diff
@@ -224,6 +230,14 @@ export async function updateArtifacts({
       PIP_CACHE_DIR: await ensureCacheDir('pip'),
     };
 
+    if (config.minimumReleaseAge) {
+      const ageMs = toMs(config.minimumReleaseAge);
+      if (isNumber(ageMs)) {
+        const days = Math.ceil(ageMs / 86_400_000);
```
Comment (RahulGautamSingh) [inline lib/modules/manager/poetry/artifacts.ts]: Sorry for this, our best-practices suggest using `luxon`, my old suggesiton was partly wrong, we should not use the magic number here:

try:
```
import { Duration } from 'luxon';               

const days = Math.ceil(Duration.fromMillis(ageMs).as('days')); 
```

File: lib/modules/manager/poetry/artifacts.ts [hunk 6]
```diff
@@ -224,6 +230,14 @@ export async function updateArtifacts({
       PIP_CACHE_DIR: await ensureCacheDir('pip'),
     };
 
+    if (config.minimumReleaseAge) {
+      const ageMs = toMs(config.minimumReleaseAge);
```
Comment (RahulGautamSingh) [inline lib/modules/manager/poetry/artifacts.ts:235]: When `toMs(config.minimumReleaseAge)` returns null, the code falls through with no log.

Add a log for this case: 
```
 logger.debug(                                                                                                                                          
    { minimumReleaseAge: config.minimumReleaseAge },
    'Invalid minimumReleaseAge, skipping POETRY_SOLVER_MIN_RELEASE_AGE',
  );
```

**Review by RahulGautamSingh — CHANGES_REQUESTED** [review]

**Review by RahulGautamSingh — COMMENTED** [review]
failing tests

File: lib/modules/manager/poetry/artifacts.ts [hunk 7]
```diff
@@ -224,6 +231,19 @@ export async function updateArtifacts({
...
+      if (isNumber(ageMs)) {
+        const days = Math.ceil(Duration.fromMillis(ageMs).as('days'));
+        extraEnv.POETRY_SOLVER_MIN_RELEASE_AGE = String(days);
```
Comment (viceice) [inline lib/modules/manager/poetry/artifacts.ts]:
```suggestion
        extraEnv.POETRY_SOLVER_MIN_RELEASE_AGE = days.toString();
```

`String` constructor is creating a wrapper object

**Comment by GodVenn** [comment]
Would love to see this feature implemented - any way I can help?

**Review by RahulGautamSingh — APPROVED** [review]

**Comment by LuigiCerone** [comment]
Hello, can we merge this? This is a really interesting feature for us, if the PR is missing something I can fix it. Thanks!

**Review by jamietanna — COMMENTED** [review]
LGTM once conflicts are resolved

**Review by RahulGautamSingh — COMMENTED** [review]
@LuigiCerone branch in conflicted, please resolve that..we can add this to merge queue then

**Review by jamietanna — APPROVED** [review]
